### ISSUE TITLE
lsp_signature with nvim cmp

### ISSUE THREAD

**Issue by Trid-collab** [ray-x/lsp_signature.nvim#93]
I was using signature with nvim compe earlier successful albeit bit slow. Now I have migrated to nvim cmp with luasnips and auto pairs - and can't seem to get the lsp signature to work. I have my settings as same as before. Now I am getting the luasnips when I opening the function brackets.

**Comment by ray-x** [comment]
It should work. I also migrated to cmp last week. Please refer to my setup.
https://github.com/ray-x/nvim

**Comment by sQVe** [comment]
I'm still having issues getting it to work. The signature window is floating above the completions. See:

![image](https://user-images.githubusercontent.com/2284724/132299332-aa34d342-c6f9-417d-b13a-150e43c9033d.png)

I'm running the following settings:

```lua
--  ┏━┓╻┏━╸┏┓╻┏━┓╺┳╸╻ ╻┏━┓┏━╸
--  ┗━┓┃┃╺┓┃┗┫┣━┫ ┃ ┃ ┃┣┳┛┣╸
--  ┗━┛╹┗━┛╹ ╹╹ ╹ ╹ ┗━┛╹┗╸┗━╸
return function()
    require'lsp_signature'.on_attach({
        bind = true,
        floating_window = true,
        handler_opts = {border = "single"},
        hint_enable = false
    })
end
```

```lua
--  ┏━╸┏━┓┏┳┓┏━┓┏━╸
--  ┃  ┃ ┃┃┃┃┣━┛┣╸
--  ┗━╸┗━┛╹ ╹╹  ┗━╸
return function()
    local cmp = require('cmp')

    local expand_snippet = function(args)
        vim.fn["vsnip#anonymous"](args.body)
    end

    local check_back_space = function()
        local col = vim.fn.col('.') - 1
        return col == 0 or vim.fn.getline('.'):sub(col, col):match('%s')
    end

    local T = function(str)
        return vim.api.nvim_replace_termcodes(str, true, true, true)
    end

    local tab_complete = function(fallback)
        if vim.fn.pumvisible() == 1 then
            vim.fn.feedkeys(T('<C-n>'), 'n')
        elseif vim.fn['vsnip#available'](1) == 1 then
            vim.fn.feedkeys(T('<Plug>(vsnip-expand-or-jump)'), '')
        elseif check_back_space() then
            vim.fn.feedkeys(T('<Tab>'), 'n')
        else
            fallback()
        end
    end

    local s_tab_complete = function(fallback)
        if vim.fn.pumvisible() == 1 then
            vim.fn.feedkeys(('<C-p>'), 'n')
        elseif vim.fn['vsnip#jumpable'](-1) == 1 then
            vim.fn.feedkeys(T('<Plug>(vsnip-jump-prev)'), '')
        else
            fallback()
        end
    end

    local get_all_buffers = function() return vim.api.nvim_list_bufs() end

    cmp.setup {
        snippet = {expand = expand_snippet},
        mapping = {
            ['<C-Space>'] = cmp.mapping.complete(),
            ['<C-e>'] = cmp.mapping.close(),
            ['<CR>'] = cmp.mapping.confirm({
                behavior = cmp.ConfirmBehavior.Replace,
                select = true
            }),
            ['<S-Tab>'] = cmp.mapping(s_tab_complete, {"i", "s"}),
            ['<Tab>'] = cmp.mapping(tab_complete, {"i", "s"})
        },
        sources = {
            {
                name = 'buffer',
                opts = {
                    get_bufnrs = get_all_buffers,
                    keyword_pattern = [[\k\+]] -- Include special characters in word match.
                }

            }, {name = 'path'}, {name = 'nvim_lsp'}, {name = 'vsnip'}
        }
    }
end
```

**Comment by ray-x** [comment]
You can set either
```
floating_window_above_cur_line = true
```


-- OR reset zindex
```
zindex = 10
```

**Comment by Trid-collab** [comment]
I am getting this 

![image](https://user-images.githubusercontent.com/61322708/132803688-4179ddd6-aae3-41a4-93e6-f34047bd817b.png)

**Comment by ray-x** [comment]
pd signature is huge... do you have enough space to display it?

![image](https://user-images.githubusercontent.com/1681295/132807507-edd0d9a0-200a-4d6b-b4c1-d0d21bd44c1f.png)

**Comment by Trid-collab** [comment]
The setting mentioned above used to work with nvim compe. it stopped when i move to cmp

I have changed it to the following settings now even then it does'nt show

cfg = function()
	require("lsp_signature").setup({
		toggle_key = nil,
		floating_window = true,
		-- floating_window_above_first = false,
		log_path = vim.fn.expand("$HOME") .. "/tmp/sig.log",
		debug = true,
		hi_parameter = "Search",
		bind = true,
		extra_trigger_chars = { "(", "," },
		handler_opts = {
			border = "single", -- "shadow", --{"╭", "─" ,"╮", "│", "╯", "─", "╰", "│" },
		},
	})
end

require("lsp_signature").on_attach(cfg)

**Comment by ray-x** [comment]
```vim
set termguicolors

call plug#begin('~/.vim/plugged')
Plug 'neovim/nvim-lspconfig'
Plug 'nvim-treesitter/nvim-treesitter'
Plug 'liuchengxu/vim-clap' ", { 'do': { -> clap#installer#force_download() } }
Plug 'jose-elias-alvarez/null-ls.nvim'
Plug 'hrsh7th/nvim-cmp'
Plug 'hrsh7th/cmp-buffer'
Plug 'hrsh7th/cmp-nvim-lsp'

" Plug 'rafamadriz/friendly-snippets'
" Plug 'hrsh7th/cmp-nvim-lua'
" Plug 'nvim-lua/popup.nvim'
" Plug 'nvim-lua/plenary.nvim'
" Plug 'nvim-telescope/telescope.nvim'

Plug '~/github/lsp_signature.nvim'
Plug '~/github/guihua.lua'

Plug '~/github/navigator.lua'
" Plug '~/github/go.nvim'

" Plug 'ray-x/guihua.lua'
" Plug 'ray-x/navigator.lua'
Plug 'ray-x/aurora'

call plug#end()



lua <<EOF

require'nvim-treesitter.configs'.setup {
  highlight = {
  enable = true,              
  additional_vim_regex_highlighting = false,
  },
}
local nvim_lsp = require('lspconfig')

-- Use an on_attach function to only map the following keys
-- after the language server attaches to the current buffer


require'lspconfig'.gopls.setup({
  flags = {
      debounce_text_changes = 150,
    }
})

local cmp = require('cmp')
cmp.setup {
   sources = {{name = 'nvim_lsp'}, {name = 'buffer'}, {name = 'nvim_lua'}}
}


require'lsp_signature'.setup({toggle_key = nil,
floating_window = true,
-- floating_window_above_first = false,
log_path = vim.fn.expand("$HOME") .. "/tmp/sig.log",
debug = true,
hi_parameter = "Search",
bind = true,
extra_trigger_chars = { "(", "," },
handler_opts = {
border = "single", -- "shadow", --{"╭", "─" ,"╮", "│", "╯", "─", "╰", "│" },
}})

EOF

set mouse=a

```

**Comment by LongerHV** [comment]
I think this issue is caused by luasnip (most likely these 'def method' snippets in python).

**Comment by ray-x** [comment]
If the luasnip enabled. should still work

![image](https://user-images.githubusercontent.com/1681295/132923556-648138b0-9c45-417e-b03f-1850e3a45430.png)

**Comment by Trid-collab** [comment]
@LongerHV  you are right. 
After a lot of trial and error with my vimrc I found that the issue arises when I have the following luasnip lines in my vimrc

require("luasnip/loaders/from_vscode").load()

but without these lines the luasnips are not showing with nvim cmp
Any indicators how to use this.

**Comment by ray-x** [comment]
As shown in my screen shot,  `def(class method)` is from luasnip.  I enabled signature, cmp, luasnip by default.

You can refer to my setup
https://github.com/ray-x/nvim/blob/master/lua/modules/completion/plugins.lua

Or  a minium vimrc:

https://github.com/ray-x/nvim/blob/master/min/vimrc.min

**Comment by ray-x** [comment]
I think might be something releated to lua snip lazyload I am using and causing failed to reproduce the issue.  A fix uploaded.

**Comment by Trid-collab** [comment]
Thanks for being so patient on this one. 
The present status is as follows. I have tried the different combinations of the "floating window, floating_window_above_cur_line ,hint_enable, fix_pos and the zindex" and even removing all the config and just loading the default. The situation is as follows:

I am not able to get the signature window on the top

![image](https://user-images.githubusercontent.com/61322708/132980230-4386d271-e151-49d4-8900-cff8d284637d.png)



local cfg = {
	bind = true,
	toggle_key = nil,
	floating_window = true,
	floating_window_above_cur_line = true,
	hint_enable = true,
	fix_pos = false,
	-- floating_window_above_first = true,
	log_path = vim.fn.expand("$HOME") .. "/tmp/sig.log",
	debug = true,
	hi_parameter = "Search",
	zindex = 200,
	timer_interval = 100,
	extra_trigger_chars = {},
	handler_opts = {
		border = "single", -- "shadow", --{"╭", "─" ,"╮", "│", "╯", "─", "╰", "│" },
	},
}

require("lsp_signature").on_attach(cfg)

**Comment by ray-x** [comment]
With your signature config setup,

The behavior should depend on your signature and current line position.
As the signature window height is 7 (5lines of signature+2lines of border)
If your current line is  > 8 the signature should look like this


![image](https://user-images.githubusercontent.com/1681295/132983234-63c67435-fcbd-4d0a-8edc-27eefd74d3d8.png)

If your current line is <9 your signature should look like this:

<img width="484" alt="Screen Shot 2021-09-12 at 7 55 02 pm" src="https://user-images.githubusercontent.com/1681295/132983270-ef31d89c-197a-4069-b78f-5f7ffbfdb304.png">


BTW, you can use
`lua print(vim.inspect(_LSP_SIG_CFG))`
print out your config and check your setup. I think z-index setup did not take effect.

**Comment by Trid-collab** [comment]
Thanks for the help. Let me check up the config using lua print you mentioned.
In the meanwhile I noticed I was using neovim 0.5 stable - after the last post I upgraded to the neovim nightly the latest version.
Now I am getting  as follows in the screenshot
It seems to working as per the patch you pushed last night. It seems also something do with the nightly and stable release as well. 
 
Thank you- this is still workable - better than nothing at all. Would be better if the lunsip option didn't show up at all underneath as shown in the screenshot.
I have opened an issue in the Luasnip forum as well

![image](https://user-images.githubusercontent.com/61322708/132989248-4ff4aff4-e6f1-4258-b8c0-d9cabfd364a4.png)

my config 
local cfg = {
	bind = true,
	toggle_key = nil,
	floating_window = true,
	floating_window_above_cur_line = true,
	hint_enable = true,
	fix_pos = false,
	-- floating_window_above_first = true,
	log_path = vim.fn.expand("$HOME") .. "/tmp/sig.log",
	debug = true,
	hi_parameter = "Search",
	zindex = 200,
	timer_interval = 100,
	extra_trigger_chars = {},
	handler_opts = {
		border = "single", -- "shadow", --{"╭", "─" ,"╮", "│", "╯", "─", "╰", "│" },
	},
}

require("lsp_signature").on_attach(cfg)
